ICS-2 Implement Misbehavior

AdityaSripal opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

AdityaSripal commented on x/ibc/02-client/types/tendermint/misbehaviour.go:
```diff
@@ -42,7 +42,5 @@ func (m Misbehaviour) ValidateBasic() error {
 
 // CheckMisbehaviour checks if the evidence provided is a misbehaviour
 func CheckMisbehaviour(m Misbehaviour) error {
```
I believe I made all the checks on misbehaviour in `Evidence.ValidateBasic` that was required in spec except for:

`abortTransactionUnless(clientState.pastPublicKeys.contains(h1.publicKey))`

From my understanding this is making sure that the validatorSets signing the Headers are known validator sets by the ClientState and were the real ValidatorSets for the height at which the evidence is submitted. Currently there is no such check, any random ValidatorSet that signs two different commit can be submitted as evidence against a chain even if that validatorSet was never a valid ValidatorSet for the chain we submit evidence against.

To fix this, the `clientState` has to know what the past `ValidatorSets` of a chain were since creation to the present, along with knowing when they were active. This information isn't currently saved in consensus state.

I think if it is saved, we can implement this check. Can I get an ACK that this functionality is needed in `ConsensusState` and my understanding of the above check is also correct before I go ahead and implement this?

The check would then proceed as:

1. Retrieve ValidatorSet for height `ev.Header1.Height` from ConsensusState
2. Check the commits in evidence against this ValidatorSet

cc: @cwgoes

cwgoes replied:
See above comments. This check is not relevant for Tendermint light clients, but other checks are.

Reviewing this PR has convinced me that the spec needs to be way more explicit here. https://github.com/cosmos/ics/issues/311

AdityaSripal commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -51,41 +51,53 @@ func (ev Evidence) Hash() cmn.HexBytes {
 
 // ValidateBasic implements Evidence interface
 func (ev Evidence) ValidateBasic() error {
```
I believe this does all checks here:

```typescript

// misbehaviour verification function defined by the client type
// any duplicate signature by a past or current key freezes the client
function checkMisbehaviourAndUpdateState(
  clientState: ClientState,
  evidence: Evidence) {
    h1 = evidence.h1
    h2 = evidence.h2
    abortTransactionUnless(clientState.pastPublicKeys.contains(h1.publicKey))
    abortTransactionUnless(h1.sequence === h2.sequence)
    abortTransactionUnless(h1.commitmentRoot !== h2.commitmentRoot || h1.publicKey !== h2.publicKey)
    abortTransactionUnless(h1.publicKey.verify(h1.signature))
    abortTransactionUnless(h2.publicKey.verify(h2.signature))
    clientState.frozen = true
}
```

except for this check:
```typescript
    abortTransactionUnless(clientState.pastPublicKeys.contains(h1.publicKey))
```

Need reviewers to double-check that this claim is true

cwgoes replied:
Hmm - this was an example in the spec, not the actual check for Tendermint.

We want to freeze the client iff. there are two valid headers at the same height.

@melekes Is this logic written out somewhere?

melekes replied:
Not yet. There's a fork accountability spec https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/fork-accountability.md.

> We want to freeze the client iff. there are two valid headers at the same height.

yeah. so basically we stop the client if we discovered (either by cross-checking the headers OR receiving evidence) that the light client is under attack. But even then, we can probably try to continue with an alternative provider.

fedekunze commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -51,41 +50,54 @@ func (ev Evidence) Hash() cmn.HexBytes {
-}
+	// Ensure that validator sets that voted on differing headers are the same validators
+	// even if headers have different proposers
+	// Require Validators are sorted first
```
are we enforcing this somehow?

marbar3778 replied:
are you sorting based in power? we briefly discussed this on the tm call on Monday but were wondering how it would affect the sdk

fedekunze commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -51,41 +50,54 @@ func (ev Evidence) Hash() cmn.HexBytes {
+	valSet1 := ev.Header1.ValidatorSet.Validators
+	valSet2 := ev.Header2.ValidatorSet.Validators
+	if len(valSet1) != len(valSet2) {
+		return errors.ErrInvalidEvidence(errors.DefaultCodespace, fmt.Sprintf("ValidatorSets have different lengths: valSet1: %s, valSet2: %s",
```
nbd but shouldn't we just print the lengths instead of the full list of validators?

fedekunze commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -51,41 +50,54 @@ func (ev Evidence) Hash() cmn.HexBytes {
+	}
+	for i, v := range valSet1 {
+		if !reflect.DeepEqual(v, valSet2[i]) {
+			return errors.ErrInvalidEvidence(errors.DefaultCodespace, fmt.Sprintf("ValidatorSets are different at index %d. v1: %s, v2: %s", i, *v, *valSet2[i]))
```
```suggestion
			return errors.ErrInvalidEvidence(errors.DefaultCodespace, fmt.Sprintf("validator sets are different at index %d. v1: %s, v2: %s", i, *v, *valSet2[i]))
```

fedekunze commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -51,41 +51,53 @@ func (ev Evidence) Hash() cmn.HexBytes {
-	return ev.DuplicateVoteEvidence.Height()
-}
+	// Convert commits to vote-sets given the validator set so we can check if they both have 2/3 power
+	voteSet1 := tmtypes.CommitToVoteSet(ev.ChainID, ev.Header1.Commit, ev.Header1.ValidatorSet)
```
cc: @melekes @ebuchman

melekes replied:
It must be a valset which created this commit. So `ev.Header1.ValidatorSet` seems right to me.

cwgoes commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -51,41 +50,54 @@ func (ev Evidence) Hash() cmn.HexBytes {
-func (ev Evidence) GetConsensusAddress() sdk.ConsAddress {
-	return sdk.ConsAddress(ev.DuplicateVoteEvidence.Address())
-}
+	// Ensure that validator sets that voted on differing headers are the same validators
```
Why would we need to check this? They might be different validators (e.g. if bisection happened).

cwgoes commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -51,41 +51,53 @@ func (ev Evidence) Hash() cmn.HexBytes {
-	return ev.DuplicateVoteEvidence.Height()
-}
+	// Convert commits to vote-sets given the validator set so we can check if they both have 2/3 power
+	voteSet1 := tmtypes.CommitToVoteSet(ev.ChainID, ev.Header1.Commit, ev.Header1.ValidatorSet)
```
Yup. Ideally we can separate this out into a helper function (Tendermint might already have one).

fedekunze commented:
@AdityaSripal can you update the `misbehaviour_test.go` file too with the new evidence?

melekes commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -19,10 +20,9 @@ var _ evidenceexported.Evidence = Evidence{}
-	ChainID        string `json:"chain_id" yaml:"chain_id"`
-	ValidatorPower int64  `json:"val_power" yaml:"val_power"`
-	TotalPower     int64  `json:"total_power" yaml:"total_power"`
+	Header1 Header
```
Maybe rename it to `SignedHeaderPlusValidatorSet` to better reflect the content?

AdityaSripal commented:
Switched Base branch to remove conflicts and make it slightly easier to review

TODO:

- [ ]  Evidence Tests
- [ ]  Misbehaviour Tests
- [ ]  Keeper.CheckMisbehaviour Tests

fedekunze commented:
@AdityaSripal should we close #5305 then?

AdityaSripal commented:
@fedekunze yea once i get the tests written, i think we can merge this straight into `ibc-alpha`

cwgoes commented:
@AdityaSripal Is this awaiting another review or did you want to finish more tests first?

fedekunze commented on x/ibc/02-client/exported/exported.go:
```diff
@@ -20,45 +17,33 @@ type ConsensusState interface {
 }
 
+// Committer defines the type that is responsible for
+// updating the consensusState at a given height
```
Can you add a note that in tendermint this is equivalent to the validator set?

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour.go:
```diff
@@ -1,96 +1,61 @@
+	trustedValSet := trustedCommitter.ValidatorSet
+	if err := trustedValSet.VerifyFutureCommit(m.Evidence.Header1.ValidatorSet, m.Evidence.ChainID,
+		m.Evidence.Header1.Commit.BlockID, m.Evidence.Header1.Height, m.Evidence.Header1.Commit); err != nil {
+		return err
```
ditto on wrapping the error with the header number and small description

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour.go:
```diff
@@ -1,96 +1,61 @@
-		return sdk.ConvertError(errors.ErrInvalidEvidence(errors.DefaultCodespace, fmt.Sprintf("Invalid Total Power: %d", ev.TotalPower)))
+	if err := trustedValSet.VerifyFutureCommit(m.Evidence.Header2.ValidatorSet, m.Evidence.ChainID,
+		m.Evidence.Header2.Commit.BlockID, m.Evidence.Header2.Height, m.Evidence.Header2.Commit); err != nil {
+		return err
```
ditto

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour_test.go:
```diff
@@ -1,64 +1,168 @@
+			err := misbehaviour.ValidateBasic()
+
+			if tc.expErr {
+				require.NotNil(suite.T(), err, "Invalid Misbehaviour passed ValidateBasic")
```
```suggestion
				suite.Error(err, "Invalid Misbehaviour passed ValidateBasic")
```

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour_test.go:
```diff
@@ -1,64 +1,168 @@
+			if tc.expErr {
+				require.NotNil(suite.T(), err, "Invalid Misbehaviour passed ValidateBasic")
+			} else {
+				require.Nil(suite.T(), err, "Valid Misbehaviour failed ValidateBasic: %v", err)
```
```suggestion
				suite.NoError(err, "Valid Misbehaviour failed ValidateBasic")
```

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour_test.go:
```diff
@@ -1,64 +1,168 @@
+			err := CheckMisbehaviour(tc.committer, misbehaviour)
+
+			if tc.expErr {
+				require.NotNil(suite.T(), err, "CheckMisbehaviour passed unexpectedly")
```
```suggestion
				suite.Error(err, "CheckMisbehaviour passed unexpectedly")
```

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour_test.go:
```diff
@@ -1,64 +1,168 @@
+			if tc.expErr {
+				require.NotNil(suite.T(), err, "CheckMisbehaviour passed unexpectedly")
+			} else {
+				require.Nil(suite.T(), err, "CheckMisbehaviour failed unexpectedly: %v", err)
```
```suggestion
				suite.NoError(err, "CheckMisbehaviour failed unexpectedly")
```

fedekunze commented on x/ibc/02-client/types/tendermint/evidence.go:
```diff
@@ -0,0 +1,101 @@
+
+func ValidCommit(chainID string, commit *tmtypes.Commit, valSet *tmtypes.ValidatorSet) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
```
and if so, why are we returning an error instead of sticking with the panic?

AdityaSripal replied:
The `CommitToVoteSet` function can panic. This makes sense in the context of how tendermint uses it, we should return error instead here since it means that someone tried to submit Evidence that wasn't actually committed by validators

This should result in just rejecting the evidence gracefully, not crashing the node

fedekunze replied:
Makes sense. Mind adding a comment on the code with that clarification?

cwgoes commented:
See https://github.com/cosmos/ics/issues/311#issuecomment-558128501 

I am working on a more formal spec document, we will discuss this in the call today.

cwgoes commented on x/ibc/02-client/exported/exported.go:
```diff
@@ -20,45 +17,33 @@ type ConsensusState interface {
 	GetRoot() commitment.RootI
 
+	// GetCommitter returns the committer that committed the consensus state
+	GetCommitter() Committer
```
I'm also confused about this method.

cwgoes commented on x/ibc/02-client/types/tendermint/misbehaviour.go:
```diff
@@ -1,96 +1,86 @@
-}
+	// Evidence is on same height as trustedCommiter. ValidatorSets must be the same
+	if trustedCommitter.GetHeight() == uint64(m.GetHeight()) {
+		if !bytes.Equal(trustedValSet.Hash(), m.Evidence.Header1.ValidatorSet.Hash()) ||
```
Why is evidence at the same height as `trustedCommiter`?

AdityaSripal replied:
This was an attempt at adding a stricter check on the validatorSets that sign off on misbehaviour that happens to be at the same height (or next height) of a known client update.

My thinking was that if the evidence was far away from the previous update. (e.g. Evidence is on height 30, previous update was on height 20), then Ill have to do the 2/3 power checking thing with the previous validatorset we know.

But if the evidence gets submitted at the same height or the next height as the previous update, i thought we could enforce a tighter bound given that we know both the validatorSet that committed the original update as well as the next validatorSet (since we persist its hash)

However, I don't believe this stricter check is correct anymore since the original update could have been made with an incorrect validator set.

So im going to remove these two conditions

cwgoes commented on x/ibc/02-client/types/tendermint/misbehaviour.go:
```diff
@@ -1,96 +1,86 @@
-}
+	// Evidence is on next height of trustedCommitter. Evidence Validator Set must match
+	// trustedCommitter's NextValSetHash
+	if trustedCommitter.GetHeight()+1 == uint64(m.GetHeight()) {
```
Why is evidence at the same height as `trustedCommiter`?

AdityaSripal commented:
Seems like there's some very understandable confusion here, given i haven't documented my thinking anywhere yet. Hopefully this comment will clarify things.

I realized that we have to persist the validatorSets on each `UpdateClient` in order to validate Evidence.

Consider the case where we updated the client at height 20 and then again at height 40

We then want to submit Evidence of misbehaviour at height 30. We should check that the validatorSets of the Evidence are within (2/3) similarity to the validatorSet we had in the latest update before 30: Height 20.
However, the current consensusState only has the validatorSet of height 40, so we need to persist the validatorSets at each update so that we can still access them and use them to check validity of Evidence. This technique of maintaining both the latest state and the past validatorSets is very similar to how it works in `tendermint/state` package.

So I needed a way to store the validatorSets in the keeper, however the keeper looks as though it should be generalized outside of tendermint-specific consensus so I introduced a new interface called `Committer` to persist validatorSets in the keeper store on each update. For now it has only the methods it needs to build, but i could add more methods to make it clearer what structs should be implementing this interface.

cwgoes commented:
> However, the current consensusState only has the validatorSet of height 40, so we need to persist the validatorSets at each update so that we can still access them and use them to check validity of Evidence. This technique of maintaining both the latest state and the past validatorSets is very similar to how it works in tendermint/state package.

Yes, precisely. This is what #311 is trying to outline. Hopefully I'll get to the spec doc soon (but sounds like you already have an idea of what needs to be done here).

I don't have a strong opinion on interfaces but keep in mind that we will need to serialise the `consensusState` (*not* past validator sets, just the current one & commit hash) in a canonical form for cross-chain verification during the handshake.

fedekunze commented on x/ibc/02-client/client/cli/tx.go:
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/cosmos/cosmos-sdk/version"
 	"github.com/cosmos/cosmos-sdk/x/auth"
 	"github.com/cosmos/cosmos-sdk/x/auth/client/utils"
+	evidence "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```
```suggestion
	evidenceexported "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```

fedekunze commented on x/ibc/02-client/client/cli/tx.go:
```diff
@@ -133,7 +134,7 @@ $ %s tx ibc client misbehaviour [client-id] [path/to/evidence.json] --from node0
 			clientID := args[0]
 
-			var evidence exported.Evidence
+			var evidence evidence.Evidence
```
```suggestion
			var evidence evidenceexported.Evidence
```

fedekunze commented on x/ibc/02-client/client/rest/rest.go:
```diff
@@ -5,6 +5,7 @@ import (
 
 	"github.com/cosmos/cosmos-sdk/client/context"
 	"github.com/cosmos/cosmos-sdk/types/rest"
+	evidence "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```
```suggestion
	evidenceexported "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```

fedekunze commented on x/ibc/02-client/client/rest/rest.go:
```diff
@@ -35,5 +36,5 @@ type UpdateClientReq struct {
 type SubmitMisbehaviourReq struct {
 	BaseReq  rest.BaseReq      `json:"base_req" yaml:"base_req"`
-	Evidence exported.Evidence `json:"evidence" yaml:"evidence"`
+	Evidence evidence.Evidence `json:"evidence" yaml:"evidence"`
```
```suggestion
	Evidence evidenceexported.Evidence `json:"evidence" yaml:"evidence"`
```

fedekunze commented on x/ibc/02-client/keeper/client.go:
```diff
@@ -66,30 +69,55 @@ func (k Keeper) UpdateClient(ctx sdk.Context, clientID string, header exported.H
+	if !found {
+		return errors.ErrCommitterNotFound(k.codespace, fmt.Sprintf("committer not found for height %d", misbehaviour.GetHeight()))
+	}
+	tmCommitter, ok := committer.(tendermint.Committer)
```
```suggestion
	
	tmCommitter, ok := committer.(tendermint.Committer)
```

fedekunze commented on x/ibc/02-client/keeper/keeper.go:
```diff
@@ -119,6 +118,32 @@ func (k Keeper) SetVerifiedRoot(ctx sdk.Context, clientID string, height uint64,
+
+	var committer exported.Committer
+
+	// TODO: Replace this for-loop with a ReverseIterator for efficiency
```
Can you create an issue for this TODO?

fedekunze commented on x/ibc/02-client/keeper/keeper_test.go:
```diff
@@ -100,3 +103,46 @@ func (suite *KeeperTestSuite) TestSetVerifiedRoot() {
+		recv, ok := suite.keeper.GetCommitter(suite.ctx, "gaia", uint64(i))
+		tmRecv, _ := recv.(tendermint.Committer)
+		if tmRecv.ValidatorSet != nil {
+			tmRecv.ValidatorSet.TotalVotingPower()
```
```suggestion
			// update validator set's power
			tmRecv.ValidatorSet.TotalVotingPower()
```

fedekunze commented on x/ibc/02-client/keeper/keeper_test.go:
```diff
@@ -100,3 +103,46 @@ func (suite *KeeperTestSuite) TestSetVerifiedRoot() {
+	suite.keeper.SetCommitter(suite.ctx, "gaia", 3, committer)
+	suite.keeper.SetCommitter(suite.ctx, "gaia", 6, nextCommitter)
+
+	for i := 0; i < 3; i++ {
```
```suggestion
	// fetch the commiter on each respective height
	for i := 0; i < 3; i++ {
```

fedekunze commented on x/ibc/02-client/keeper/keeper_test.go:
```diff
@@ -100,3 +103,46 @@ func (suite *KeeperTestSuite) TestSetVerifiedRoot() {
+		recv, ok := suite.keeper.GetCommitter(suite.ctx, "gaia", uint64(i))
+		tmRecv, _ := recv.(tendermint.Committer)
+		if tmRecv.ValidatorSet != nil {
+			tmRecv.ValidatorSet.TotalVotingPower()
```
```suggestion
			// update validator set's power
			tmRecv.ValidatorSet.TotalVotingPower()
```

fedekunze commented on x/ibc/02-client/types/msgs.go:
```diff
@@ -2,12 +2,20 @@ package types
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	evidence "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```
```suggestion
	evidenceexported "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```

fedekunze commented on x/ibc/02-client/types/msgs.go:
```diff
@@ -120,12 +128,12 @@ func (msg MsgUpdateClient) GetSigners() []sdk.AccAddress {
 type MsgSubmitMisbehaviour struct {
 	ClientID string            `json:"id" yaml:"id"`
-	Evidence exported.Evidence `json:"evidence" yaml:"evidence"`
+	Evidence evidence.Evidence `json:"evidence" yaml:"evidence"`
```
```suggestion
	Evidence evidenceexported.Evidence `json:"evidence" yaml:"evidence"`
```

fedekunze commented on x/ibc/02-client/types/msgs.go:
```diff
@@ -120,12 +128,12 @@ func (msg MsgUpdateClient) GetSigners() []sdk.AccAddress {
 
 // NewMsgSubmitMisbehaviour creates a new MsgSubmitMisbehaviour instance
-func NewMsgSubmitMisbehaviour(id string, evidence exported.Evidence, signer sdk.AccAddress) MsgSubmitMisbehaviour {
+func NewMsgSubmitMisbehaviour(id string, evidence evidence.Evidence, signer sdk.AccAddress) MsgSubmitMisbehaviour {
```
```suggestion
func NewMsgSubmitMisbehaviour(id string, evidence evidenceexported.Evidence, signer sdk.AccAddress) MsgSubmitMisbehaviour {
```

fedekunze commented on x/ibc/02-client/types/msgs_test.go:
```diff
@@ -9,11 +9,38 @@ import (
 	cmn "github.com/tendermint/tendermint/libs/common"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	evidence "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```
```suggestion
	evidenceexported "github.com/cosmos/cosmos-sdk/x/evidence/exported"
```

fedekunze commented on x/ibc/02-client/types/msgs_test.go:
```diff
@@ -9,11 +9,38 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types/tendermint"
 )
 
+var _ evidence.Evidence = mockEvidence{}
```
```suggestion
var _ evidenceexported.Evidence = mockEvidence{}
```

fedekunze commented on x/ibc/02-client/types/msgs_test.go:
```diff
@@ -9,11 +9,38 @@ import (
 )
 
+var _ evidence.Evidence = mockEvidence{}
+var _ evidence.Evidence = mockBadEvidence{}
```
```suggestion
var _ evidenceexported.Evidence = mockBadEvidence{}
```

fedekunze commented on x/ibc/02-client/types/tendermint/committer.go:
```diff
@@ -0,0 +1,23 @@
+
+var _ exported.Committer = Committer{}
+
+type Committer struct {
```
missing godoc comments and `json` and `yaml` tags

fedekunze commented on x/ibc/02-client/types/tendermint/consensus_state.go:
```diff
@@ -38,6 +39,14 @@ func (cs ConsensusState) GetRoot() commitment.RootI {
 	return cs.Root
 }
 
+func (cs ConsensusState) GetCommitter() exported.Committer {
```
missing godoc

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour.go:
```diff
@@ -3,94 +3,64 @@ package tendermint
-	ChainID        string `json:"chain_id" yaml:"chain_id"`
-	ValidatorPower int64  `json:"val_power" yaml:"val_power"`
-	TotalPower     int64  `json:"total_power" yaml:"total_power"`
+// Misbehaviour contains an evidence that a
```
```suggestion
// Misbehaviour contains evidence that a light client submitted a different header from
// a full node at the same height.
```

fedekunze commented on x/ibc/02-client/types/tendermint/misbehaviour.go:
```diff
@@ -3,94 +3,64 @@ package tendermint
-func (ev Evidence) GetHeight() int64 {
-	return ev.DuplicateVoteEvidence.Height()
-}
+// CheckMisbehaviour checks if the evidence provided is a misbehaviour
```
```suggestion
// CheckMisbehaviour checks if the evidence provided is a valid light client misbehaviour
```

AdityaSripal commented:
> With regards to UX, is the Commiter something that we want to expose through the CLI and REST for querying purposes? Asking bc the client Misbehaviour is already queryable through the evidence module. This can be done in a separate PR if necessary.

Would prefer doing this in a separate PR if possible, this PR is already getting pretty lengthy and that functionality isn't strictly necessary yet
